File: Napakalaki/src/napakalaki/BadConsequence.java

package napakalaki;

import static java.lang.Math.min;
import java.util.ArrayList;

/**
 *
 * @author alex
 */
public class BadConsequence {
    
    // DECLARACIÓN DE VARIABLES
    
    private String text;
    private int levels;
    private int nVisibleTreasures;
    private int nHiddenTreasures;
    private boolean death;
    static int MAXTREASURES = 10;
    
    private ArrayList<TreasureKind> specificHiddenTreasures;
    private ArrayList<TreasureKind> specificVisibleTreasures;

    // CONSTRUCTORES

    public BadConsequence(String text, int levels, int nVisibleTreasures, int nHiddenTreasures) {
        this.text = text;
        this.levels = levels;
        this.nVisibleTreasures = nVisibleTreasures;
        this.nHiddenTreasures = nHiddenTreasures;
        death = false;
        specificVisibleTreasures = new ArrayList();
        specificHiddenTreasures = new ArrayList();
    }

    public BadConsequence(String text, boolean death) {
        this.text = text;
        levels = 0;
        nVisibleTreasures = 0;
        nHiddenTreasures = 0;
        this.death = death;
        specificVisibleTreasures = new ArrayList();
        specificHiddenTreasures = new ArrayList();
    }

    public BadConsequence(String text, int levels, ArrayList<TreasureKind> tVisible,
            ArrayList<TreasureKind> tHidden) {
        this.text = text;
        this.levels = levels;
        nVisibleTreasures = 0;
        nHiddenTreasures = 0;
        death = false;
        specificVisibleTreasures = tVisible;
        specificHiddenTreasures = tHidden;
    }


    // GETTERS
    
    public String getText() {
        return text;
    }

    public int getLevels() {
        return levels;
    }

    public int getnVisibleTreasures() {
        return nVisibleTreasures;
    }

    public int getnHiddenTreasures() {
        return nHiddenTreasures;
    }

    public boolean isDeath() {
        return death;
    }

    public ArrayList<TreasureKind> getSpecificHiddenTreasures() {
        return specificHiddenTreasures;
    }

    public ArrayList<TreasureKind> getSpecificVisibleTreasures() {
        return specificVisibleTreasures;
    }
    
    // MÉTODOS
    
    
    // Devuelve true si no hay mal rollo por cumplir
    
    public boolean isEmpty(){
        return levels==0 && death==false && nVisibleTreasures==0 && nHiddenTreasures==0 &&
               specificVisibleTreasures.isEmpty() && specificHiddenTreasures.isEmpty();
    }
    
    
    // Elimina el tesoro t del mal rollo.
    
    public void substractVisibleTreasure(Treasure t){
        if(!specificVisibleTreasures.isEmpty())
            specificVisibleTreasures.remove(t.getType());
        else if(nVisibleTreasures != 0)
            nVisibleTreasures--;
    }
       
    // Elimina el tesoro t del mal rollo.
    
    public void substractHiddenTreasure(Treasure t){
        if(!specificHiddenTreasures.isEmpty())
            specificHiddenTreasures.remove(t.getType());
        else if(nHiddenTreasures != 0)
            nHiddenTreasures--;
    }
    
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h){
        BadConsequence adjustedBC;
        
        if(specificVisibleTreasures.isEmpty() && specificHiddenTreasures.isEmpty()){
            int adjustedVisible;
            int adjustedHidden;
            
            if(v.size() < nVisibleTreasures)
                adjustedVisible = v.size();
            else
                adjustedVisible = nVisibleTreasures;
            if(h.size() < nHiddenTreasures)
                adjustedHidden = h.size();
            else
                adjustedHidden = nHiddenTreasures;
            
            adjustedBC = new BadConsequence(text, 0, adjustedVisible, adjustedHidden);
        }
        else{
            ArrayList<TreasureKind> adjustedSpecificVisible = new ArrayList();
            ArrayList<TreasureKind> adjustedSpecificHidden = new ArrayList();

            
            for(TreasureKind tKind : TreasureKind.values()) {
                    int min1 = 0, min2 = 0;
                    
                    // Trabajamos con tesoros visibles para cada TreasureKind
                       
                    for(int i=0; i<specificVisibleTreasures.size(); i++){
                        if(specificVisibleTreasures.get(i) == tKind)
                            min1++; 
                    }
                    for(int i=0; i<v.size(); i++){
                        if(v.get(i).getType() == tKind)
                            min2++; 
                    }
                    // Tomamos el mínimo y añadimos ese número de TreasureKind
                    for(int i=0; i < min(min1, min2);  i++){
                        adjustedSpecificVisible.add(tKind); 
                    }
                        
                    //Reseteamos los contadores
                    min1 = min2 = 0; 
                        
                    // Trabajamos con tesoros ocultos. 
                    for(int  i=0; i<specificHiddenTreasures.size(); i++){
                        if(specificHiddenTreasures.get(i) == tKind)
                            min1++; 
                    }
                    for(int i=0; i<h.size(); i++){
                        if(h.get(i).getType() == tKind)
                            min2++; 
                    }
                    // Tomamos el mínimo y añadimos ese número de TreasureKind
                    for(int i = 0; i < min(min1, min2);  i++){
                        adjustedSpecificHidden.add(tKind); 
                    }
            }  
            
            adjustedBC = new BadConsequence(text, 0, adjustedSpecificVisible, adjustedSpecificHidden);         
        }

        return adjustedBC;
    }
    
    // Método to_string
    
    @Override
    public String toString() {
        return  text + ", niveles=" + levels + 
                ", Tesoros Visibles=" + nVisibleTreasures + ", Tesoros Ocultos="
                + nHiddenTreasures + ", Muerte=" + death + ", Tesoros Visibles Espec.="
                + specificVisibleTreasures + ", Tesoros Ocultos Espec.=" + specificHiddenTreasures;
    }
    
}

File: NapakalakiRuby/pkg/NapakalakiRuby-0.0.1/lib/bad_consequence.rb
# coding: utf-8

require_relative 'treasure_kind.rb'

module Napakalaki

  class BadConsequence
  
    # CONSULTORES
  
    #attr_reader :text, :levels, :nVisibleTreasures, :nHiddenTreasures, :death,
     # :specificVisibleTreasures, :specificHiddenTreasures, :MAXTREASURES
     
    def getLevels
      @levels
    end
    
    def isDeath
      @death
    end
    
    # INICIALIZADOR
    
    @@MAXTREASURES=10
    def initialize(text, levels, nVisibleTreasures, nHiddenTreasures, specificVisibleTreasures,
        specificHiddenTreasures, death)
      @text=text
      @levels=levels
      @nVisibleTreasures=nVisibleTreasures
      @nHiddenTreasures=nHiddenTreasures
      @specificVisibleTreasures=specificVisibleTreasures
      @specificHiddenTreasures=specificHiddenTreasures
      @death=death
    end
  
    private_class_method :new;
  
    def self.newLevelNumberOfTreasures(txt, lvl, visibleTreasures, hiddenTreasures)
      new(txt, lvl, visibleTreasures, hiddenTreasures, Array.new, Array.new, false)
    end
  
    def self.newLevelSpecificTreasures(txt, lvl, specVisibleTreasures, specHiddenTreasures)
      new(txt, lvl, 0, 0, specVisibleTreasures, specHiddenTreasures, false)
    end
  
    def self.newDeath(txt)
      new(txt, 10, @@MAXTREASURES, @@MAXTREASURES, Array.new, Array.new, true)
    end
    
    # MÉTODOS
    
    # Devuelve true si no hay mal rollo por cumplir.
    
    def isEmpty
      empty=false
      
      if(@levels==0 and @death==false and @nVisibleTreasures==0 and @nHiddenTreasures==0 and 
         @specificVisibleTreasures.empty? and @specificHiddenTreasures.empty? )
        empty=true
      end
      
      return empty
    end
    
    
    # Elimina el tesoro t del mal rollo.
    
    def substractVisibleTreasure(t)
      if(!@specificVisibleTreasures.empty?)
        index=@specificVisibleTreasures.index(t.getType)
        @specificVisibleTreasures.delete_at(index)
      elsif(@nVisibleTreasures != 0)
          @nVisibleTreasures -= 1
      end
    end
    
    
    # Elimina el tesoro t del mal rollo.
    
    def substractHiddenTreasure(t)
      if(!@specificHiddenTreasures.empty?)
        index=@specificHiddenTreasures.index(t.getType)
        @specificHiddenTreasures.delete_at(index)
      elsif(@nHiddenTreasures != 0)
          @nHiddenTreasures -= 1
      end
    end
    
    def adjustToFitTreasureLists(v, h)
      if(@specificVisibleTreasures.empty? and @specificHiddenTreasures.empty?)
        nVTreasures = [v.size, @nVisibleTreasures].min
        nHTreasures = [h.size, @nHiddenTreasures].min

        bc = BadConsequence.newLevelNumberOfTreasures(@text, 0, nVTreasures, nHTreasures)
      else
        adjusted_specific_visible=Array.new
        adjusted_specific_hidden=Array.new
        treasure_kind=Array.new
        
        treasure_kind << TreasureKind::ARMOR
        treasure_kind << TreasureKind::ONEHAND
        treasure_kind << TreasureKind::BOTHHANDS
        treasure_kind << TreasureKind::HELMET
        treasure_kind << TreasureKind::SHOES
        puts "treasure kind loop"
        treasure_kind.each do |tKind|
          min1=min2=0

          # Contamos cuantos tesoros tenemos de cada tipo en el jugador y en BC
          
          @specificVisibleTreasures.each do |vKind|
            if(vKind == tKind)
              min1+=1
            end
          end
          puts "sv loop"
          v.each do |vKind|
            if(vKind.getType == tKind)
              min2+=1
            end
          end
          puts "v loops"
          # Añadimos el numero mínimo de tesoros al BC
          puts min1
          puts min2
          i=0
          while(i<[min1, min2].min)
            adjusted_specific_visible << tKind
            i+=1
          end
          puts "add v"
          # Reseteamos contadores y hacemos lo mismo para los tesoros ocultos
          
          min1=min2=0
          
          @specificHiddenTreasures.each do |hKind|
            if(hKind == tKind)
              min1+=1
            end
          end
          puts "sh loop"
          h.each do |hKind|
            if(hKind.getType == tKind)
              min2+=1
            end
          end
          puts "h loops"
          # Añadimos el numero mínimo de tesoros al BC
          puts min1
          puts min2
          i=0
          while(i<[min1, min2].min)
            adjusted_specific_hidden << tKind
            i+=1
          end
          puts "add h"
        end
        
        bc = BadConsequence.newLevelSpecificTreasures(@text, 0, adjusted_specific_visible, adjusted_specific_hidden)
        
      end
        puts bc.to_s
      return bc
      
    end
    
    # Método to_string
    
    def to_s
      "Mal Rollo-> #{@text}, niveles: #{@levels},
        Tesoros Visibles: #{@nVisibleTreasures}, Tesoros Ocultos: #{@nHiddenTreasures},
        Muerte: #{@death}, Tesoros Visibles Espec.: #{@specificVisibleTreasures}, 
        Tesoros Ocultos Esp.: #{@specificHiddenTreasures}";
    end
  
  end

end